File: Python/match_to_fruns.py
"""Match franchise names to FRUNS identifiers.

Matches input franchise names against the FRUNS master file using
sanitization, harmonization, and optional fuzzy matching.
"""

from pathlib import Path
from typing import Literal

import polars as pl
from rapidfuzz import process
from rapidfuzz.distance import JaroWinkler

from normalize_franchise_names import sanitize_name, harmonize_name


# Data loading -----------------------------------------------------------------


def load_fruns_data() -> pl.DataFrame:
    """Load FRUNS master data from CSV."""
    data_path = Path(__file__).parent.parent / "data" / "fruns-master.csv"
    return pl.read_csv(
        data_path,
        schema_overrides={
            "fruns": pl.Utf8,
            "last_avail_year": pl.Int64,
            "brand_name": pl.Utf8,
            "brand_name_sanitized": pl.Utf8,
            "franchisor": pl.Utf8,
            "franchisor_sanitized": pl.Utf8,
            "naics_code": pl.Utf8,
            "naics_description": pl.Utf8,
        },
    )


def load_harmonize_map() -> pl.DataFrame:
    """Load harmonization mappings from CSV."""
    data_path = Path(__file__).parent.parent / "data" / "harmonize-names.csv"
    return pl.read_csv(
        data_path,
        schema_overrides={"franchise": pl.Utf8, "name_harmonized": pl.Utf8},
        null_values="NA",
    )


# Main function ----------------------------------------------------------------


def match_to_fruns(
    data: pl.DataFrame,
    name_col: str,
    method: Literal["both", "exact", "fuzzy"] = "both",
    fruns_data: pl.DataFrame | None = None,
    harmonize_map: pl.DataFrame | None = None,
    max_distance: float = 0.10,
    verbose: bool = True,
    keep_details: bool = False,
) -> pl.DataFrame:
    """Match franchise names to FRUNS identifiers.

    Args:
        data: DataFrame containing franchise names to match.
        name_col: Column name containing franchise names.
        method: Matching method:
            - "exact": exact matching only (fast, no false positives)
            - "fuzzy": fuzzy matching only (for testing/diagnostics)
            - "both": exact first, then fuzzy for unmatched (recommended)
        fruns_data: FRUNS lookup table (defaults to loading from data/).
        harmonize_map: Harmonization mappings (defaults to loading from data/).
        max_distance: Maximum Jaro-Winkler distance for fuzzy matches (default 0.10).
        verbose: Print match summary and sample fuzzy matches.
        keep_details: Keep diagnostic columns (name_sanitized, match_key,
            fruns_name_sanitized, distance).

    Returns:
        Input data with 'fruns' and 'match_type' columns added.
    """
    if fruns_data is None:
        fruns_data = load_fruns_data()
    if harmonize_map is None:
        harmonize_map = load_harmonize_map()

    prepared = _prepare_input(data, name_col, harmonize_map)

    if method == "exact":
        exact_matches = _find_exact_matches(prepared, fruns_data)
        result = _finalize_matches(prepared, exact_matches, fruns_data, keep_details)
        if verbose:
            _print_summary(result)
        return result

    if method == "fuzzy":
        fuzzy_matches = _match_fuzzy(prepared, fruns_data, max_distance)
        result = _finalize_matches(prepared, fuzzy_matches, fruns_data, keep_details)
        if verbose:
            _print_summary(result, fuzzy_matches)
        return result

    # method == "both": exact first, then fuzzy for unmatched
    exact_matches = _find_exact_matches(prepared, fruns_data)

    matched_ids = set(exact_matches["_row_id"].to_list())
    unmatched = prepared.filter(~pl.col("_row_id").is_in(matched_ids))

    if len(unmatched) == 0:
        fuzzy_matches = pl.DataFrame()
    else:
        fuzzy_matches = _match_fuzzy(unmatched, fruns_data, max_distance)

    all_matches = pl.concat([exact_matches, fuzzy_matches], how="diagonal_relaxed")
    result = _finalize_matches(prepared, all_matches, fruns_data, keep_details)
    if verbose:
        _print_summary(result, fuzzy_matches)
    return result


def _prepare_input(
    data: pl.DataFrame, name_col: str, harmonize_map: pl.DataFrame
) -> pl.DataFrame:
    """Step 1 & 2: Sanitize and harmonize input names."""
    prepared = data.with_row_index("_row_id")
    sanitized = sanitize_name(prepared[name_col])
    harmonized = harmonize_name(sanitized, harmonize_map)
    return prepared.with_columns(
        sanitized.alias("_name_sanitized"),
        harmonized.alias("_match_key"),
    )


def _find_exact_matches(
    data: pl.DataFrame, fruns_data: pl.DataFrame
) -> pl.DataFrame:
    """Step 3a & 3b: Find exact matches on brand and franchisor names."""
    # Filter out rows with null/empty harmonized names
    data_valid = data.filter(
        pl.col("_match_key").is_not_null()
        & (pl.col("_match_key") != "")
    )

    # Step 3a: Exact match on brand name
    fruns_brand = (
        fruns_data.select("fruns", "brand_name_sanitized")
        .drop_nulls(subset=["brand_name_sanitized"])
        .filter(pl.col("brand_name_sanitized") != "")
    )

    # Validate m:1 - brand names should be unique in FRUNS
    assert fruns_brand["brand_name_sanitized"].is_unique().all(), (
        "brand_name_sanitized is not unique in FRUNS data"
    )

    exact_brand = data_valid.join(
        fruns_brand,
        left_on="_match_key",
        right_on="brand_name_sanitized",
        how="inner",
    ).with_columns(pl.lit("exact").alias("match_type"))

    # Step 3b: Exact match on franchisor name (for rows not matched by brand)
    matched_ids = set(exact_brand["_row_id"].to_list())
    unmatched = data_valid.filter(~pl.col("_row_id").is_in(matched_ids))

    fruns_franchisor = (
        fruns_data.select("fruns", "franchisor_sanitized")
        .drop_nulls(subset=["franchisor_sanitized"])
        .filter(pl.col("franchisor_sanitized") != "")
    )

    # Many-to-many join to detect multiple matches
    franchisor_joined = unmatched.join(
        fruns_franchisor,
        left_on="_match_key",
        right_on="franchisor_sanitized",
        how="inner",
    )

    if len(franchisor_joined) == 0:
        return exact_brand

    # Count matches per row
    match_counts = (
        franchisor_joined.group_by("_row_id")
        .len()
        .rename({"len": "_n_matches"})
    )
    franchisor_joined = franchisor_joined.join(match_counts, on="_row_id", how="left")

    # Single franchisor matches - unambiguous
    exact_franchisor = (
        franchisor_joined.filter(pl.col("_n_matches") == 1)
        .with_columns(pl.lit("franchisor").alias("match_type"))
        .drop("_n_matches")
    )

    # Multiple franchisor matches - ambiguous, return null with special match_type
    multiple_ids = (
        franchisor_joined.filter(pl.col("_n_matches") > 1)
        .select("_row_id")
        .unique()
    )
    franchisor_multiple = (
        unmatched.join(multiple_ids, on="_row_id", how="inner")
        .with_columns(
            pl.lit(None).cast(pl.Utf8).alias("fruns"),
            pl.lit("franchisor_multiple").alias("match_type"),
        )
    )

    return pl.concat(
        [exact_brand, exact_franchisor, franchisor_multiple],
        how="diagonal_relaxed",
    )


def _match_fuzzy(
    data: pl.DataFrame, fruns_data: pl.DataFrame, max_distance: float
) -> pl.DataFrame:
    """Step 3c: Find best fuzzy match via Jaro-Winkler distance.

    Uses rapidfuzz.process.extractOne for O(n) performance with C-optimized
    string comparison, replacing the previous O(n*m) nested Python loop.
    """
    unique_input = (
        data["_match_key"].drop_nulls().unique().to_list()
    )
    unique_brands = (
        fruns_data["brand_name_sanitized"].drop_nulls().unique().to_list()
    )

    # Filter empty strings
    unique_input = [name for name in unique_input if name]
    unique_brands = [brand for brand in unique_brands if brand]

    if not unique_input or not unique_brands:
        return pl.DataFrame()

    # Convert max_distance to minimum similarity score (extractOne uses similarity)
    score_cutoff = 1 - max_distance

    # Find best match for each unique input name using optimized extractOne
    best_matches = []
    for name in unique_input:
        result = process.extractOne(
            name,
            unique_brands,
            scorer=JaroWinkler.similarity,
            score_cutoff=score_cutoff,
            processor=None,  # Names are already sanitized
        )

        if result is not None:
            best_brand, score, _ = result
            distance = 1 - score

            # Round to 4 decimal places for consistent comparison with R
            if round(distance, 4) <= max_distance:
                best_matches.append(
                    {
                        "_match_key": name,
                        "_matched_brand": best_brand,
                        "_distance": distance,
                    }
                )

    if not best_matches:
        return pl.DataFrame()

    matches_df = pl.DataFrame(best_matches)

    # Join to get FRUNS
    fruns_lookup = (
        fruns_data.select("brand_name_sanitized", "fruns")
        .unique(subset=["brand_name_sanitized"])
    )

    # Validate m:1
    assert fruns_lookup["brand_name_sanitized"].is_unique().all()

    matches_df = matches_df.join(
        fruns_lookup,
        left_on="_matched_brand",
        right_on="brand_name_sanitized",
        how="left",
    )

    # Join back to original data
    result = data.join(matches_df, on="_match_key", how="inner")
    result = result.with_columns(
        pl.col("_distance")
        .map_elements(lambda d: f"fuzzy_{d:.4f}", return_dtype=pl.Utf8)
        .alias("match_type")
    )

    return result


def _finalize_matches(
    prepared: pl.DataFrame,
    matches: pl.DataFrame,
    fruns_data: pl.DataFrame,
    keep_details: bool = False,
) -> pl.DataFrame:
    """Join matches back to original data."""
    if keep_details:
        keep_cols = ["_row_id", "fruns", "match_type", "_distance"]
        available_cols = [c for c in keep_cols if c in matches.columns]
        matches_subset = (
            matches.select(available_cols) if len(matches) > 0
            else pl.DataFrame(schema={c: pl.Utf8 for c in keep_cols})
        )

        result = prepared.join(matches_subset, on="_row_id", how="left")

        # Canonical sanitized brand name of the matched fruns - unlike the
        # match key, this is guaranteed unique per fruns for all match types
        fruns_names = fruns_data.select(
            "fruns",
            pl.col("brand_name_sanitized").alias("_fruns_name_sanitized"),
        )
        result = result.join(fruns_names, on="fruns", how="left")

        # Place fruns_name_sanitized right after match_type
        cols = result.columns
        cols.remove("_fruns_name_sanitized")
        cols.insert(cols.index("match_type") + 1, "_fruns_name_sanitized")
        result = result.select(cols)

        # Rename internal columns (remove leading underscore)
        rename_map = {c: c[1:] for c in result.columns if c.startswith("_")}
        result = result.rename(rename_map)
    else:
        cols = ["_row_id", "fruns", "match_type"]
        available_cols = [c for c in cols if c in matches.columns]
        matches_subset = (
            matches.select(available_cols) if len(matches) > 0
            else pl.DataFrame(schema={c: pl.Utf8 for c in cols})
        )

        result = prepared.join(matches_subset, on="_row_id", how="left")
        # Drop internal columns
        internal_cols = [c for c in result.columns if c.startswith("_")]
        result = result.drop(internal_cols)

    return result


def _print_summary(
    result: pl.DataFrame, fuzzy_matches: pl.DataFrame | None = None
) -> None:
    """Print match summary to console."""
    match_bucket = (
        pl.when(pl.col("match_type").is_null())
        .then(pl.lit("unmatched"))
        .when(pl.col("match_type").str.starts_with("fuzzy"))
        .then(pl.lit("fuzzy"))
        .otherwise(pl.col("match_type"))
    )

    counts = (
        result.with_columns(match_bucket.alias("match_bucket"))
        .group_by("match_bucket")
        .len()
        .rename({"len": "n"})
        .with_columns((pl.col("n") / pl.col("n").sum() * 100).alias("pct"))
    )

    total = len(result)
    matched = counts.filter(pl.col("match_bucket") != "unmatched")["n"].sum()

    print("\n-- Match Summary --")
    print(f"{total} rows -> {matched} matched ({matched/total*100:.1f}%)")
    print()

    for row in counts.iter_rows(named=True):
        print(f"  {row['match_bucket']:<12} {row['n']:>6}  {row['pct']:>5.1f}%")

    if fuzzy_matches is not None and len(fuzzy_matches) > 0:
        print("\n-- Fuzzy Match Samples --")
        sample = fuzzy_matches.sample(n=min(10, len(fuzzy_matches)))
        sample = sample.sort("_distance")

        for row in sample.iter_rows(named=True):
            name = str(row["_match_key"])[:30].ljust(30)
            brand = str(row["_matched_brand"])[:30].ljust(30)
            dist = row["_distance"]
            print(f"  {name}  ->  {brand}  {dist:.3f}")


if __name__ == "__main__":
    # Example usage
    test_data = pl.DataFrame(
        {
            "franchise_name": [
                "McDonald's Franchising, Inc.",
                "Subway",
                "Some Unknown Franchise",
                "7-Eleven",
            ]
        }
    )

    result = match_to_fruns(test_data, "franchise_name", keep_details=True)
    print("\n-- Result --")
    print(result)

File: R/match_to_fruns.R
#' Match franchise names to FRUNS identifiers
#'
#' Matches input franchise names against the FRUNS master file using
#' sanitization, harmonization, and optional fuzzy matching.
#'
#' @param data Data frame containing franchise names to match
#' @param name_col Column containing franchise names (unquoted)
#' @param method Matching method:
#'   - "exact": exact matching only (fast, no false positives)
#'   - "fuzzy": fuzzy matching only (for testing/diagnostics)
#'   - "both": exact first, then fuzzy for unmatched (recommended)
#' @param fruns_data FRUNS lookup table (defaults to package data)
#' @param harmonize_map Harmonization mappings (defaults to package data)
#' @param max_distance Maximum Jaro-Winkler distance for fuzzy matches (default 0.10)
#' @param verbose Print match summary and sample fuzzy matches (default TRUE)
#' @param keep_details Keep diagnostic columns: name_sanitized, name_harmonized, distance (default FALSE)
#' @return Input data with `fruns` and `match_type` columns added

source("R/normalize_franchise_names.R")

# Data loading -----------------------------------------------------------------

load_fruns_data <- function() {
  readr::read_csv("data/fruns-master.csv", show_col_types = FALSE)
}

load_harmonize_map <- function() {
  # Keep NA mappings - they're intentional exclusions for generic names
  # (e.g., "pizza", "plumber") that would otherwise cause false positives.
  readr::read_csv("data/harmonize-names.csv", show_col_types = FALSE)
}

# Main function ----------------------------------------------------------------

match_to_fruns <- function(
  data,
  name_col,
  method = c("both", "exact", "fuzzy"),
  fruns_data = NULL,
  harmonize_map = NULL,
  max_distance = 0.10,
  verbose = TRUE,
  keep_details = FALSE
) {
  method <- match.arg(method)

  fruns_data <- fruns_data %||% load_fruns_data()
  harmonize_map <- harmonize_map %||% load_harmonize_map()

  prepared <- prepare_input(data, {{ name_col }}, harmonize_map)

  if (method == "exact") {
    exact_matches <- find_exact_matches(prepared, fruns_data)
    result <- finalize_matches(prepared, exact_matches, keep_details)
    if (verbose) {
      print_summary(result)
    }
    return(result)
  }

  if (method == "fuzzy") {
    fuzzy_matches <- match_fuzzy(prepared, fruns_data, max_distance)
    result <- finalize_matches(prepared, fuzzy_matches, keep_details)
    if (verbose) {
      print_summary(result, fuzzy_matches)
    }
    return(result)
  }

  # method == "both": exact first, then fuzzy for unmatched
  exact_matches <- find_exact_matches(prepared, fruns_data)

  unmatched <- prepared |>
    dplyr::anti_join(exact_matches, by = dplyr::join_by(.row_id))

  if (nrow(unmatched) == 0) {
    fuzzy_matches <- dplyr::tibble()
  } else {
    fuzzy_matches <- match_fuzzy(unmatched, fruns_data, max_distance)
  }

  all_matches <- dplyr::bind_rows(exact_matches, fuzzy_matches)
  result <- finalize_matches(prepared, all_matches, keep_details)
  if (verbose) {
    print_summary(result, fuzzy_matches)
  }
  result
}

# Step 1: Sanitization ---------------------------------------------------------
# Clean up input names: lowercase, replace @ with "at", remove apostrophes/dots,
# strip corporate suffixes (Inc, LLC, Corp), franchise terms (Franchise,
# Franchising, Franchisor), and trailing terms (International, Holdings, USA).
# Example: "McDonald's Franchising, Inc." → "mcdonalds"

# Step 2: Harmonization --------------------------------------------------------
# Apply manual mapping from harmonize-names.csv. Maps known variants to
# canonical names (e.g., "mcd" → "mcdonalds"). NA mappings intentionally block
# generic terms from matching (e.g., "food" → NA prevents false positives).

# Step 3: Matching -------------------------------------------------------------
# Find the FRUNS ID using the harmonized name:
#   - Exact (brand): match against brand_name_sanitized in FRUNS master
#   - Exact (franchisor): if no brand match, try franchisor_sanitized
#   - Fuzzy: if still unmatched, find closest brand via Jaro-Winkler (≤0.10)

# Helpers ----------------------------------------------------------------------

prepare_input <- function(data, name_col, harmonize_map) {
  data |>
    dplyr::mutate(
      .row_id = dplyr::row_number(),
      .name_sanitized = sanitize_name({{ name_col }}),
      .name_harmonized = harmonize_name(.name_sanitized, harmonize_map)
    )
}

find_exact_matches <- function(data, fruns_data) {
  # Step 3a: Exact match on brand name
  # Brand names should be unique in FRUNS - error if multiple matches found
  exact_brand <- data |>
    dplyr::inner_join(
      fruns_data |> dplyr::select(fruns, brand_name_sanitized),
      by = dplyr::join_by(.name_harmonized == brand_name_sanitized),
      na_matches = "never",
      relationship = "many-to-one"
    ) |>
    dplyr::mutate(match_type = "exact")

  # Step 3b: Exact match on franchisor name (for rows not matched by brand)
  # Multiple franchises may share a franchisor name - handle explicitly
  unmatched_by_brand <- data |>
    dplyr::anti_join(exact_brand, by = dplyr::join_by(.row_id))

  franchisor_joined <- unmatched_by_brand |>
    dplyr::inner_join(
      fruns_data |> dplyr::select(fruns, franchisor_sanitized),
      by = dplyr::join_by(.name_harmonized == franchisor_sanitized),
      na_matches = "never",
      relationship = "many-to-many"
    ) |>
    dplyr::mutate(
      .n_matches = dplyr::n(),
      .by = .row_id
    )

  # Single franchisor matches
  exact_franchisor <- franchisor_joined |>
    dplyr::filter(.n_matches == 1) |>
    dplyr::select(-.n_matches) |>
    dplyr::mutate(match_type = "franchisor")

  # Multiple franchisor matches - ambiguous, cannot determine correct FRUNS
  franchisor_multiple <- franchisor_joined |>
    dplyr::filter(.n_matches > 1) |>
    dplyr::distinct(.row_id, .keep_all = TRUE) |>
    dplyr::mutate(
      fruns = NA_character_,
      match_type = "franchisor_multiple"
    ) |>
    dplyr::select(-.n_matches)

  dplyr::bind_rows(exact_brand, exact_franchisor, franchisor_multiple)
}

finalize_matches <- function(prepared, matches, keep_details = FALSE) {
  if (keep_details) {
    keep_cols <- c(
      ".row_id",
      "fruns",
      "match_type",
      ".matched_brand",
      ".distance"
    )
    matches <- matches |>
      dplyr::select(dplyr::any_of(keep_cols))

    prepared |>
      dplyr::left_join(matches, by = dplyr::join_by(.row_id)) |>
      dplyr::rename_with(
        \(x) stringr::str_remove(x, "^\\."),
        dplyr::starts_with(".")
      )
  } else {
    matches <- matches |>
      dplyr::select(.row_id, fruns, match_type)

    prepared |>
      dplyr::left_join(matches, by = dplyr::join_by(.row_id)) |>
      dplyr::select(-dplyr::starts_with("."))
  }
}

print_summary <- function(result, fuzzy_matches = NULL) {
  counts <- result |>
    dplyr::mutate(
      match_bucket = dplyr::case_when(
        is.na(match_type) ~ "unmatched",
        startsWith(match_type, "fuzzy") ~ "fuzzy",
        .default = match_type
      )
    ) |>
    dplyr::count(match_bucket) |>
    dplyr::mutate(pct = n / sum(n) * 100)

  total <- nrow(result)
  matched <- sum(counts$n[counts$match_bucket != "unmatched"])

  cli::cli_h2("Match Summary")
  cli::cli_text(
    "{.strong {total}} rows \u2192 {.strong {matched}} matched ({round(matched/total*100, 1)}%)"
  )

  # Tabular summary
  max_type <- max(nchar(counts$match_bucket))
  fmt <- paste0("  %-", max_type, "s  %6d  %5.1f%%")
  cli::cli_verbatim("")
  cli::cli_verbatim(sprintf(fmt, counts$match_bucket, counts$n, counts$pct))

  if (!is.null(fuzzy_matches) && nrow(fuzzy_matches) > 0) {
    samples <- fuzzy_matches |>
      dplyr::slice_sample(n = min(10, nrow(fuzzy_matches))) |>
      dplyr::arrange(.distance)

    cli::cli_h2("Fuzzy Match Samples")
    cli::cli_verbatim(sprintf(
      "  %-30.30s  \u2192  %-30.30s  %.3f",
      samples$.name_harmonized,
      samples$.matched_brand,
      samples$.distance
    ))
  }

  invisible(result)
}

# Step 3c: Fuzzy matching ------------------------------------------------------

#' Find best fuzzy match for each unmatched name via Jaro-Winkler (≤0.10)
match_fuzzy <- function(data, fruns_data, max_distance) {
  unique_input <- unique(data$.name_harmonized)
  unique_brands <- unique(fruns_data$brand_name_sanitized)

  # Find best match index directly (no full matrix needed)
  best_idx <- stringdist::amatch(
    unique_input,
    unique_brands,
    method = "jw",
    p = 0.1,
    maxDist = max_distance
  )

  has_match <- !is.na(best_idx)
  if (!any(has_match)) {
    return(dplyr::tibble())
  }

  # Build matches for inputs that found something
  best_matches <- dplyr::tibble(
    .name_harmonized = unique_input[has_match],
    .matched_brand = unique_brands[best_idx[has_match]]
  ) |>
    dplyr::mutate(
      .distance = stringdist::stringdist(
        .name_harmonized,
        .matched_brand,
        method = "jw",
        p = 0.1
      )
    )

  # Join to get FRUNS and back to original data
  best_matches |>
    dplyr::left_join(
      fruns_data |> dplyr::distinct(brand_name_sanitized, fruns),
      by = dplyr::join_by(.matched_brand == brand_name_sanitized)
    ) |>
    dplyr::inner_join(data, by = dplyr::join_by(.name_harmonized)) |>
    dplyr::mutate(match_type = paste0("fuzzy_", round(.distance, 4)))
}
